Autocomplete links to user posts

BoardJames opened the pull request:
## Description
This implements autocompletion of user names as in #2793 . Currently it is only implemented on the paragraph block though that could be extended.
To try it out:

1. Press '@' anywhere in the paragraph where there is a preceding space or the text is interrupted somehow.
2. A menu listing all the users will be shown
3. Start typing a user name to filter the list and user the UP/DOWN arrow keys to change the selected item.
4. Press enter or click an item in the list to select it
5. A link to the user posts will be inserted replacing the \@query 

## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
This has been tested with automated tests and manually.

## Screenshots (jpeg or gifs if applicable):
**After pressing '@':**
![user-autocomplete-1](https://user-images.githubusercontent.com/26243021/31265975-fdee6de4-aab4-11e7-9a79-3e7af20e0ed5.png)
**After typing 'jam' to filter:**
![user-autocomplete-2](https://user-images.githubusercontent.com/26243021/31265983-05f46372-aab5-11e7-860b-dfce9614ee3e.png)
**After pressing enter:**
![user-autocomplete-3](https://user-images.githubusercontent.com/26243021/31265989-0aceacb8-aab5-11e7-8b13-eb5f3fefbacd.png)

## Types of changes
This is mostly a new feature built on top of the existing block-autocomplete and autocomplete features. I had to heavily modify the autocomplete feature to make it generic enough to work for user autocompletes. I also had to modify the popover code slightly to allow it to position over a node range.
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

tiny-james committed:
Autocomplete links to user posts

youknowriad commented on blocks/library/paragraph/index.js:
```diff
@@ -152,39 +153,41 @@ registerBlockType( 'core/paragraph', {
-					onReplace={ onReplace }
-					placeholder={ placeholder || __( 'New Paragraph' ) }
+			<UserAutocomplete key="editable">
+				<BlockAutocomplete onReplace={ onReplace }>
```
I haven't looked in details yet, but I'm wondering if we should create a separate Component for this or enhance the BlockAutocomplete component instead. I guess will have `tags` autocomplete too, so we may end-up with several levels of autocomplete.

Trying to think about what the API should look like, maybe I see something like this:

```js
const blockPattern = {
  triggerOn: ( content ) => content.length === '', // when to trigger the autocomplete could use regex for instance
  getSuggestions: ( content ) => fetch( '/users' ), // A function triggers on each content change if the the triggerOn is true, it returns the list of suggestions (or a promise of list of suggestions)
  renderSuggestion: ( suggestion ) => <Button>{ suggestion.tittle }</Button> // Renders a unique suggestion
  onSelect: ( suggestion ) => // What happens when you select a suggestion
}

const patterns =[ blockPattern, userPattern, tagsPattern ];

<Autocomplete patterns={ patterns } />
```

And the patterns could be declared in separate files for reusability.

Thoughts @aduth

EphoxJames replied:
The API you suggest is insufficient. If you want it to work for triggers like '@' anywhere in the content then you also need to know the location of the cursor otherwise you will popup dialogs for items that match the trigger but you are nowhere near.

EphoxJames replied:
Also for things like the `@user` popups you only want them to happen outside of links because a link inside a link is invalid HTML so you have to work with the actual HTML not just the text.

EphoxJames replied:
Anyway, I am working on making it work like you suggest though I won't have it finished today.

youknowriad replied:
Yes, I'm aware the API is probably not feature-proof but just wanted to show an idea on how we can build this generically. (which would ease extending these for plugins as well).

EphoxJames replied:
I have updated the design of Autocomplete to take in multiple "completers".

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
add/2793: Changed autocomplete to take in "completer" definitions

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
add/2793: Improved inline documentation

tiny-james committed:
add/2793: Removed old autocomplete components

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
add/2793: Updating tests [WIP]

tiny-james committed:
add/2793: Updated the rest of the existing tests

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

EphoxJames commented:
@youknowriad If you have another chance please look over the changes and advise if you think it needs any more work or if you're happy.

youknowriad commented on blocks/autocompleters/index.js:
```diff
@@ -0,0 +1,132 @@
+		className: 'blocks-block-autocomplete',
+		triggerPrefix: '/',
+		getOptions,
+		allowContext,
```
Can't understand the difference between `triggerPrefix` and  `allowContext`, could these be merged into one function?

EphoxJames replied:
TLDR: not without making it an annoying API to use.

triggerPrefix is the string that preceeds the query. For example "@" in the case of the user mentions, "/" in the case of block selection and "#" in the case of hash-tags (which aren't implemented yet). In theory you could just search back to the last space or boundary and pass the whole range to the completer definition to check but that that means that ALL completers would have to convert the range into text and check the leading substring. This puts more work on the implementer of the completer which is exactly where we don't want it.

allowContext was added to solve the problem that sometimes the completer has to know what preceeds it and follows it. The majority of the time this doesn't matter but the original implementation of the autocomplete block would require that there was nothing but whitespace before and afterwards (which makes sense because it replaces the whole block). This field is completely optional and most completers won't use it.

youknowriad commented on blocks/autocompleters/index.js:
```diff
@@ -0,0 +1,132 @@
+		mention.textContent = '@' + user.name;
+		range.insertNode( mention );
+		range.setStartAfter( mention );
+		range.deleteContents();
```
I wonder if the `userAutocompleter` should be aware of what to do with the selected value. This forces us to add a `range` property here where we could just execute a TinyMCE command if we were to have an instance of it. 

Same could be said above for the blocks autocompleter, assuming there's only one action to be done when selecting a block: replacing.

What if the `onSelect` is always provided by the calling side:

```js
<Autocomplete key="editable" completers={ [
	blockAutocompleter( block => onReplace( block ) ),
	userAutocompleter( user => this.editor.insertContent( linkFromUser( user ) ) ),
] }>
```

youknowriad replied:
I guess the difficulty here is that we need to remove what we already typed. Maybe we should do this consistently before calling the `onSelect`

EphoxJames replied:
The advantage of Range is that it is an editor agnostic API and in the past I have been told I can't assume that components will always work with TinyMCE. I specifically did not want the autocomplete code to make any assumptions about what would be done with the range.
Note that TinyMCE doesn't really care about us changing the content out from under it and it would be breaking the abstraction that Editable is suppose to provide.

EphoxJames replied:
Also one of the advantages of using range is that TinyMCE knows how to work with ranges.
For example if TinyMCE was available as `this.editor` then you could do:
```javascript
(user, range) => {
  this.editor.selection.setRng( range );
  this.editor.insertContent( linkFromUser( user ) );
}
```

youknowriad commented on blocks/library/paragraph/index.js:
```diff
@@ -152,7 +152,10 @@ registerBlockType( 'core/paragraph', {
+			<Autocomplete key="editable" completers={ [
+				blockAutocompleter( { onReplace } ),
+				userAutocompleter(),
+			] }>
```
Nicee 👍

youknowriad replied:
Thinking the "patterns" we have right now in `Editable` could be written like this cc @iseulde

youknowriad commented on components/autocomplete/index.js:
```diff
@@ -3,11 +3,12 @@
  */
```
I'm not comfortable reviewing the changes in this file, I feel like it could be simplified if we do the generic `onSelect` callback. I believe @aduth created this file originally and would love if you have time to look at these changes (maybe after the change above)

EphoxJames replied:
I am happy to explain what I've done line by line if it would help?

youknowriad reviewed (commented):
Overall, this is going to the right direction IMO. What are your thoughts about the simplification I'm proposing

tiny-james committed:
add/2793: Removed blank line that linter complained about

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
add/2793: Allow onSelect to optionally return selection replacement

tiny-james committed:
add/2793: Made linter happy again

EphoxJames commented:
@youknowriad I have tried to simplify onSelect further by allowing the provider to simply return the react html that they want to replace the range with - it will render it to html and insert it into the page and remove the autocomplete trigger and query. That way authors of autocompleters don't have to even know what a range is but it is available if they know and want to do something more complex.

EphoxJames commented on components/autocomplete/index.js:
```diff
@@ -34,86 +99,179 @@ class Autocomplete extends Component {
 		this.setState( this.constructor.getInitialState() );
 	}
 
-	onBlur( event ) {
```
Explanation:
onBlur was removed because the aim of hiding the popup when clicking away could be better handled by using the onClickOutside function of Popover.

aduth replied:
Problem here is if the user tabs away from the block while the autocomplete is still open, it remains open.

EphoxJames replied:
Well the original (as in the version in master) was hiding the Popover when you clicked on it meaning that options could only be selected by pressing enter. I will try to find a way to avoid both problems though.

EphoxJames replied:
I ended up adding a handler for the tab key to close the menu.

EphoxJames replied:
I realized that the left and right arrow keys had the same problem so they also close the menu now.

aduth replied:
>Well the original (as in the version in master) was hiding the Popover when you clicked on it meaning that options could only be selected by pressing enter.

Ah yeah, seeing this now on master. I can't say for certain that detecting specific keys as an implied blur is sufficient though. Do these differ at all by browser, user preferences, OS? Might be missing some like Escape.

Actually, in the case of Escape it may be working correctly, but only because [Popover stops propagation when handling this itself](https://github.com/WordPress/gutenberg/blob/ec8df21abb643cb5e67cd998269d53c0af61e654/components/popover/index.js#L198). Quite fragile.

Seems like the problem is that since the popover is rendered elsewhere in the DOM, `this.node.contains` will return `false` when clicking on a button within the popoover. But since our component here is the one responsible for rendering the list, maybe a solution is to assign a ref for the `ul` and check whether the related target is also (or instead) in this list.

Might be a bug worth addressing in a separate pull request.

EphoxJames commented on components/autocomplete/index.js:
```diff
@@ -34,86 +99,179 @@ class Autocomplete extends Component {
+		const container = event.target;
+		const cursor = this.getCursor( container );
+		// look for the trigger prefix and search query just before the cursor location
+		const match = ( function() {
```
Note:
match() does the following:
1. Search backwards from the cursor looking for the first space or element boundary (ie text highlighted bold)
2. Then matches the trigger prefix of the completers  to the text immediately after the space/element boundary.
In the process it will filter the list of completers by 2 other critera:
1. All text nodes included in the match must be considered acceptable by the completer - this helps us check that we're not trying to create a link inside another link.
2. The range before and after the match must be acceptable to the completer - this helps us ensure that we're not going to replace a block that actually has content in it.

aduth replied:
Should this just be a separate function? `getMatch` ? The IIFE might not be an obvious pattern for the uninitiated reader.

EphoxJames replied:
I extracted it out as findMatch

aduth commented on components/autocomplete/index.js:
```diff
@@ -176,40 +351,76 @@ class Autocomplete extends Component {
+	}
+
+	toggleKeyEvents( isListening ) {
+		// This exists because we must capture ENTER key presses before Editable.
```
If enter works correctly in master with the block autocomplete, why do we suddenly need this now?

EphoxJames replied:
The enter key handler in master relied on overriding the keyhandler of the Editable. I'm not certain why it was able to get the event before TinyMCE when a onKeyDownCapture handler on the surrounding component was not - I guess the react simulated events don't mix with native events perfectly. When I was initially working on nesting these autocomplete handlers that approach was unworkable though now it is only one component I suppose we could return to it. That said I think that overriding the event handler on a child block is very weird and I much prefer this approach.

aduth commented on components/autocomplete/index.js:
```diff
@@ -176,40 +351,76 @@ class Autocomplete extends Component {
+		// Some completers must do asynchronous requests to get their options so
+		// we calculate that here.
+		completers.forEach( ( { getOptions }, idx ) => {
+			getOptions().then( ( options ) => {
```
This behavior means we're issuing one request to the users endpoint for every paragraph block in a post. I'm not sure we need to be so proactive about querying this either? In some cases won't we need fresh data in response to a user search?

EphoxJames replied:
Good point - I will change it to populate the cache on the first non-empty search of a lookup.

EphoxJames replied:
Now it loads options on first use.

EphoxJames replied:
I have tweaked it so it also makes the getOptions call each time the autocompleter is initially triggered, so the information will be fresh, though query characters that narrow down the search will not cause a new getOptions call.

aduth commented on components/autocomplete/index.js:
```diff
@@ -176,40 +351,76 @@ class Autocomplete extends Component {
-					isOpen={ isOpen && filteredOptions.length > 0 }
+					isOpen={ filteredOptions.length > 0 }
 					focusOnOpen={ false }
+					onClickOutside={ () => this.reset() }
```
Should this be binding to `onClose` ? We can't assume clicking outside is the only reason for a popover to close (escape press).

EphoxJames replied:
I hadn't noticed `onClose` - I will investigate.

EphoxJames replied:
It seems that onClose is no different to onClickOutside with exception of when the Popover is handling the ESCAPE key (which it is not in this case).

aduth commented on components/autocomplete/index.js:
```diff
@@ -176,40 +351,76 @@ class Autocomplete extends Component {
+		completers.forEach( ( { getOptions }, idx ) => {
+			getOptions().then( ( options ) => {
+				this.allOptions[ idx ] = options.map( ( option, key ) => ( { ...option, key } ) );
+				this.forceUpdate();
```
Rather than forcing the update, should we be setting all options into state via `setState`?

EphoxJames replied:
Yeah I probably should. I did it this way because I was too lazy to rewrite the reset function (which would have deleted the cache the way it is currently written).

EphoxJames replied:
I have moved the options into state.

aduth commented on components/autocomplete/index.js:
```diff
@@ -19,12 +20,76 @@ import Popover from '../popover';
+			text.length - 1 :
+			Math.min( text.length - 1, fromIndex );
+	for ( let i = fromI; i >= 0; i-- ) {
+		if ( /\s/.test( text.charAt( i ) ) ) {
```
This function is called on every keypress in a paragraph block, which is definitely not ideal, and if we're to keep it, should be subject to optimizing for performance. The RegExp test is not the fastest:

https://jsperf.com/character-whitespace

aduth replied:
Maybe this could serve better than our own iterating implementation:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/search

EphoxJames replied:
Well I suppose I could use a regex like:
```javascript
/\s\S*$/.exec(text)
```
Still I have no idea if that is better than a loop searching from the end without benchmarking it and it would probably be highly dependent on the length of the text.

I should point out that:
```javascript
' \t\n\r\v'.indexOf( testChar ) !== -1
```
is NOT equivalent for example it does not match nonbreaking space.

EphoxJames replied:
I did some benchmarking and even on a bad case the existing code can run 655,000 times per second. I tried the regex that I mention above and of course which is better depends on the length of the string, the regex works better for short strings but once the string becomes long the search from the end wins. In short I don't think there is a good reason to replace this function.

aduth commented on components/autocomplete/index.js:
```diff
@@ -34,86 +99,194 @@ class Autocomplete extends Component {
+				}
+			}
+			// exit early if nothing can handle it
+			if ( text.substr( pos + 1 ).length === 0 ) {
```
Another optimization: We probably don't need to do a full `substr` if we're just testing if there's more text left, instead looking at the immediate next character:

```js
if ( text[ pos + 1 ] === undefined ) {
```

EphoxJames replied:
Or:
```javascript
if ( text.length <= pos + 1 )
```

aduth commented on components/autocomplete/index.js:
```diff
@@ -176,40 +351,76 @@ class Autocomplete extends Component {
 						{ filteredOptions.map( ( option, index ) => (
 							<li
-								key={ option.value }
+								key={ option.key.toString() }
```
What's the case where the `toString` becomes necessary?

EphoxJames replied:
According to react docs key should be a String - I wanted to use a number so I converted it to a string.

EphoxJames replied:
From this page:
https://reactjs.org/docs/lists-and-keys.html
"A “key” is a special string attribute you need to include when creating lists of elements. "

EphoxJames replied:
Anyway I have now changed this so the key is the combined indexes of the completer and the option separated by underscore so the toString is getting removed.

aduth commented on components/autocomplete/index.js:
```diff
@@ -176,40 +351,76 @@ class Autocomplete extends Component {
+					onClickOutside={ () => this.reset() }
 					position="top right"
 					className={ classes }
+					range={ range }
```
I'm going to have to take a closer look at the needs here, but at a glance it doesn't appear `range` is intuitively a concern that Popover should be handling.

EphoxJames replied:
I needed some way of positioning the Popover above a selection. Range seemed the best way to pass a selection.

EphoxJames replied:
There is probably a better name for it though - range is rather generic.

EphoxJames commented on blocks/autocompleters/index.js:
```diff
@@ -0,0 +1,128 @@
+				return {
+					value: user,
+					label: [
+						<img key="avatar" alt="" src={ user.avatar_urls[ 24 ] } />,
```
The problem with that is that screen readers will (for each item) read out "User avatar" before telling them the actually useful information of the user-name. Unless we can give an actually useful description it's probably more usable if they are just treated as decoration.

EphoxJames replied:
@afercia do you think we should put descriptive text on the avatar in the users menu or treat it as decorative?

afercia replied:
@EphoxJames empty `alt` are fine in this case. The relevant information is already in plain text close to the image, so these images are decorative.
Although not exactly the same thing, we've faced a similar case in the WP credits page. See https://core.trac.wordpress.org/changeset/36406 and https://core.trac.wordpress.org/ticket/34953.

EphoxJames commented on blocks/autocompleters/index.js:
```diff
@@ -0,0 +1,128 @@
+ */
+export function userAutocompleter() {
+	const getOptions = () => {
+		return ( new wp.api.collections.Users() ).fetch().then( ( users ) => {
```
If you know of a better source for the information I will happily switch to it.

tiny-james committed:
add/2793: extract out findMatch; load options on demand; handle tab

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
add/2793: Hide the popover on left and right arrow keys

tiny-james committed:
add/2793: Moved options into state and reload them on each new open

EphoxJames commented:
How do I fix this "No report found to compare against" error?

tiny-james committed:
add/2793: Removed unused allOptions variable

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
add/2793: changed onClickOutside to onClose

EphoxJames commented:
Since this issue has gotten a bit long here's a summary of the review comments and my response.
- [x] youknowriad suggested merging the block-autocompleter and user-autocompleter into one configurable component
- [ ] youknowriad suggested combining `triggerPrefix` and `allowContext` - my objection is that I think it would make the API much more annoying to use
- [ ] youknowriad suggested eliminating the `range` parameter to `onSelect` - my objection is that I think the range is useful, but as a compromise I changed `onSelect` to optionally return content to replace with, which meant that the range did not need to be used in the common case.
- [x] aduth sugested that the popup should be hidden when the user tabs away
- [x] aduth sugested that the assignment to match should be extracted into a separate function
- [ ] aduth noted that the users list is restricted to admins - currently I don't know of an alternative though suggestions are welcome.
- [ ] aduth suggested that the avatar in the popup should have alt text - I am worried that it might actually make things more wordy for screen readers without any gain but I have asked @afercia for advice
- [ ] aduth suggested changing the whitespace search to be more efficient - I investigated and found that it was acceptably efficient at 655,000 OPS for an unusually bad test case and that the alternatives suggested have problematic downsides.
- [x] aduth suggested optimizing the empty text check
- [ ] aduth suggested using the same key handling code as in master - my objection to making this change is that it will be fragile to changes as in master the autocompleter must be the direct parent of Editable to work.
- [x] aduth suggested lazy loading of the options and regularly refreshing the cache
- [x] aduth suggested storing the options in state
- [x] aduth suggested using `onClose` instead of `onClickOutside`
- [ ] aduth suggested that `range` should not be needed by Popover - my objection is that the Popover can't be positioned in an intuitive way otherwise, though I welcome suggestions.
- [x] aduth suggested removing the `toString` on the popover menu item key attributes

@youknowriad  and @aduth have I answered your reviews to your satisfaction?

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

afercia commented:
With the current implementation, screen readers are not able to announce anything in the list of suggestions:

![screen shot 2017-10-20 at 09 05 32](https://user-images.githubusercontent.com/1682452/31809235-fe092892-b576-11e7-9623-ec4fe8952ad1.png)

They just keep reading where the caret/focus is, that's a native behavior.

I've recommended during a recent editor team chat that this functionality needs to be implemented like the tags and links suggestions but I see the direction taken has been different. The result is this is not accessible and needs to be revisited. Precious time could have been saved implementing it in the right way since the beginning.

The only ways I can think of to make this work with screen readers are:
- use an input field and the same ARIA roles and attributes as in the links and tags suggestions: this works and we already have working examples
- try to move focus to the menu items, and test it in all the major browsers/screen readers combinations

afercia commented:
Here's what Twitter does: they use the same technique Gutenberg and WordPress use on the links and tags suggestions. aria-activedescendand ant other attributes are used directly on the whole contenteditable tweet box:

<img width="1002" alt="screen shot 2017-10-20 at 10 01 54" src="https://user-images.githubusercontent.com/1682452/31811435-d9ae3a24-b57f-11e7-8e3c-741bed9edb57.png">

afercia commented:
Testing with Safari Technology Preview (not sure why Safari 11 fails when using VoiceOver):

![screen shot 2017-10-20 at 10 37 49](https://user-images.githubusercontent.com/1682452/31812420-321e65dc-b583-11e7-833e-3db5c42dd77c.png)

EphoxJames commented:
@afercia thanks for your input I will see what I can do to make it more accessible. Unfortunately moving the focus to the menu is not an option in this case so I will have to find another way.

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

EphoxJames commented:
So as the required accessibility changes require new aria props on Editable I hope to put the framework in place with #3106 and then once it is merged make the changes required here.

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
add/2793: Updated tests to work with updated enzyme

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
Fixed linting error.

tiny-james committed:
add/2793: Try to make autocomplete popover accessable

tiny-james committed:
add/2793: Updated autocomplete tests to handle FACC structure

tiny-james committed:
add/2793: delegate handling of nil aria attributes to Editable

Currently Editable's aria attribute copying dumbly adds nil attributes
but that should be fixed in another PR.

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
add/2793: Disabled linter warning about findDOMNode as it is required

EphoxJames commented:
@afercia I have tested the changes with NVDA and they seem to work though the menu items only seem to be read when I moved through them with the arrow keys. I also tried JAWS professional but I couldn't figure out how to edit anything (I think I was stuck in navigation mode) so I couldn't test if it was working. Do you think it is ready to merge?

@aduth @youknowriad Any further issues I need to address before merging? Do you disagree with my resolution of any of the existing issues? If not can someone approve the merge?

aduth commented on components/autocomplete/index.js:
```diff
@@ -169,48 +329,87 @@ class Autocomplete extends Component {
-					isOpen={ isOpen && filteredOptions.length > 0 }
+					isOpen={ isOpen }
 					focusOnOpen={ false }
+					onClose={ () => this.reset() }
```
We should avoid creating a new function reference each time the component renders, since otherwise React will be forced to reconcile the change; this can be simply `onClose={ this.reset }`

EphoxJames replied:
Done.

aduth commented on components/popover/index.js:
```diff
@@ -133,17 +133,27 @@ class Popover extends Component {
+			if ( isLeft ) {
+				rect = rects[ 0 ];
+			} else if ( isRight ) {
+				rect = rects[ rects.length - 1 ];
```
Reading through the documentation for [`Element#getClientRects`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getClientRects), it's not clear to me what we're achieving by referencing the start/end of the returned collection?

EphoxJames replied:
For the case of auto-completion it will do nothing because all the text will be on one line however if this was used on something that could be line wrapped then positioning over the first rect for left alignment and the last rect for right alignment seems to be sensible to me.

aduth commented on blocks/library/paragraph/index.js:
```diff
@@ -151,39 +151,48 @@ registerBlockType( 'core/paragraph', {
+				blockAutocompleter( { onReplace } ),
+				userAutocompleter(),
+			] }>
+				{ ( isExpanded, listBoxId, activeId ) => (
```
Do we expect to need to enhance this with more options, and/or is the ordering important here? Wondering if we'd be better off specifying this as an object from which we can pick, so instead:

```js
{ ( { isExpanded, listBoxId, activeId } ) => (
```

EphoxJames replied:
It's unlikely that we'll need more options but I don't see a downside so sure.

aduth commented on components/autocomplete/index.js:
```diff
@@ -169,48 +329,87 @@ class Autocomplete extends Component {
 				className="components-autocomplete"
 			>
-				{ cloneElement( Children.only( children ), {
-					onInput: this.search,
```
Since we added Editable "event proxying" specifically to allow for this, and given some of the challenges we've encountered with the event proxying, I'm wondering if now that we're no longer overriding these props that we drop the Editable logic:

https://github.com/WordPress/gutenberg/blob/ce7870c821a4e80c2df27da2aa17af64dfc7eae4/blocks/editable/index.js#L146-L163

EphoxJames replied:
Fine with me but should probably be in another PR.

aduth commented on components/autocomplete/index.js:
```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+	// this method is separate so it can be overrided in tests
+	getCursor( container ) {
+		const selection = window.getSelection();
+		if ( selection.isCollapsed && container.contains( selection.anchorNode ) ) {
```
Is there any chance we'd get to this point and `container.contains( selection.anchorNode )` be false? Seems it could be fair to assume `input` event would only fire if selection is within the container? Trying to see if we can drop some logic which is run on every keystroke in Editable.

EphoxJames replied:
Ok, I have moved it into an assertion so it is only run during development.

aduth commented on components/autocomplete/index.js:
```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
-					selectedIndex: 0,
-					search,
-				} );
+		if ( endTextNode === null ) {
```
Can we expect this to occur often? Seeing as `completers` is not used to this point, we could avoid mapping `allCompleters` until _after_ we finish checking early returns, to avoid wasted logic run on every keystroke in editable.

EphoxJames replied:
I have moved the calculation of `completers`.

aduth commented on components/autocomplete/index.js:
```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+			// put the open completer at the start so it has priority
+			completers = [
+				wasOpen,
+				...completers.filter( ( completer ) => completer.idx !== wasOpen.idx ),
```
Do we need the `Array#map` and `Array#filter` ? Wondering if we could omit by the reference directly, since this is run on every keystroke in editable:

```js
let completers = allCompleters;
if ( wasOpen ) {
	completers = [
		wasOpen,
		...without( completers, wasOpen )
	];
}
```

https://lodash.com/docs/4.17.4#without

EphoxJames replied:
The 'without' method won't work here because it will use object equality which won't be true.

EphoxJames replied:
Relating to your later comment about the performance of `Array#filter` I will change this to lodash's filter.

aduth commented on components/autocomplete/index.js:
```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+		let pos = lastIndexOfSpace( text );
+		while ( pos === -1 ) {
+			const prev = onlyTextNode( startTextNode.previousSibling );
+			if ( prev ) {
```
Minor: We could do an early break to avoid indenting the rest of the `while`:

```js
if ( ! prev ) {
	break;
}

// ...
```

EphoxJames replied:
Done.

aduth commented on components/autocomplete/index.js:
```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+			const prev = onlyTextNode( startTextNode.previousSibling );
+			if ( prev ) {
+				// filter the completers to those that could handle this node
+				completers = completers.filter(
```
In this performance-critical code, I might suggest opting for Lodash equivalents to native functions, which tend to perform better:

https://jsperf.com/lodash-vs-underscore-vs-native-filter/2

EphoxJames replied:
Well lodash does seem to get more stable performance across browsers but it is only the best on Chrome. Still considering we are talking about filtering a list of 3 items worrying about this is kinda pointless.

I performed these tests on the same MacBook Air for consistency.

## Firefox

Filter | OPS
------ | ---
Native filter | 92,841 ±6.30% fastest
Lodash filter | 47,316 ±2.99% 47% slower
Underscore filter | 38,479 ±8.49% slower

## Chrome

Filter | OPS
------ | ---
Native filter | 11,803 ±2.06% 76% slower
Lodash filter | 49,174 ±1.89% fastest
Underscore filter | 35,244 ±1.94% slower

## Safari

Filter | OPS
------ | ---
Native filter | 22,662 ±9.40% 64% slower
Lodash filter | 43,801 ±12.18% 32% slower
Underscore filter | 61,778 ±7.03% fastest

EphoxJames replied:
Note I have changed everything to use lodash's filter and map.

aduth commented on components/autocomplete/index.js:
```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+			}
+			return true;
+		} );
+		if ( open ) {
```
Minor: Can flip this to do early return and avoid indentation of the next handful of lines:

```js
if ( ! open ) {
	return null;
}
```

EphoxJames replied:
Done.

aduth commented on components/autocomplete/index.js:
```diff
@@ -169,48 +329,87 @@ class Autocomplete extends Component {
+		// native browser event has already bubbled so we can't stopPropagation
+		//  and avoid Editable getting the event from TinyMCE, hence we must
+		// register a native event handler.
+		if ( ! this.node ) {
```
When is `this.node` falsey?

EphoxJames replied:
This won't be called very often so I doubt it will have much effect on performance but I have removed the guard as requested.

aduth replied:
It's less about performance or safe-guarding, more to having knowledge to when those safe-guards should be expected to be encountered.

aduth commented on components/autocomplete/index.js:
```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+			const range = this.createRange( startTextNode, pos + 1, endTextNode, endIndex );
+			const before = this.createRange( container, 0, startTextNode, pos + 1 );
+			const after = this.createRange( endTextNode, endIndex, container, container.childNodes.length );
+			if ( ! allowContext( before, range, after ) ) {
```
Where are we using `range`?

EphoxJames replied:
I couldn't think of a use case so I have removed it.

aduth commented on components/autocomplete/index.js:
```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+			const before = this.createRange( container, 0, startTextNode, pos + 1 );
+			const after = this.createRange( endTextNode, endIndex, container, container.childNodes.length );
+			if ( ! allowContext( before, range, after ) ) {
+				return false;
```
Seems like this could be simplified to `return allowContext( before, range, after );` (or cast to boolean via `!!` is we might expect non-boolean return types)

EphoxJames replied:
Done - was a vestige of the original where there was only one completer and they were nested rather than a list.

aduth commented on components/autocomplete/index.js:
```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+		const match = this.findMatch( container, cursor, completers, wasOpen );
+		const { open, query, range } = match || {};
+		// create a regular expression to filter the options
+		const search = open ? new RegExp( escapeStringRegexp( query ), 'i' ) : /./;
```
Minor and unrelated to these changes: I discovered that Lodash has a utility for this built-in, so we could probably drop the extra dependency:

https://lodash.com/docs/4.17.4#escapeRegExp

EphoxJames replied:
Ok I will switch over to the lodash version.

aduth replied:
This is the only place we were using `escape-string-regexp` so we should remove the dependency as well.

aduth commented on components/popover/index.js:
```diff
@@ -133,17 +133,27 @@ class Popover extends Component {
 		const isLeft = 'left' === xAxis;
 		const isRight = 'right' === xAxis;
-
+		let rect = anchor.parentNode.getBoundingClientRect();
```
Seems wasteful to call `Element#getBoundingClientRect` only to have chance of overriding if `range` is specified. This looks like something that could go into the `else` of the below condition:

```js
let rect;
if ( range ) {
	// ...
} else {
	rect = anchor.parentNode.getBoundingClientRect();
}
```

EphoxJames replied:
Done.

aduth commented on components/popover/index.js:
```diff
@@ -133,17 +133,27 @@ class Popover extends Component {
 		const isRight = 'right' === xAxis;
-
+		let rect = anchor.parentNode.getBoundingClientRect();
+		if ( range ) {
```
Is the general idea that if we pass a range, we use it as a anchor reference instead of `this.nodes.anchor` ? If so, do we need to do anything else about the logic below for padding offset (should this be eliminated)?

Or, to my earlier points about making Popover unaware of range, would one of the following options work:

- Allow parent component to pass an override for setting offset, like `onSetOffset` or `getOffsetRect` ?
- Allow parent component to pass an anchor object which implements `getBoundingClientRect` (`Range`, `Element`)?

EphoxJames replied:
> Is the general idea that if we pass a range, we use it as a anchor reference instead of this.nodes.anchor ?

yes

> If so, do we need to do anything else about the logic below for padding offset (should this be eliminated)?

Yes it looks like it is not needed when positioned over a range.

> Or, to my earlier points about making Popover unaware of range, would one of the following options work:
> * Allow parent component to pass an override for setting offset, like onSetOffset or getOffsetRect ?
> * Allow parent component to pass an anchor object which implements getBoundingClientRect (Range, Element)?

They'd probably both work to some degree though the second option would be problematic when ranges wrapped across multiple lines.

EphoxJames replied:
I have changed the calculation of the topOffset so it only happens for the anchor.

EphoxJames replied:
I have extracted the getBoundingClientRect logic into a function getAnchorRect which can be passed to override the default of positioning around the parent element.

So it now behaves like your first suggested option.

tiny-james committed:
add/2793: Converted autocomplete FACC parameters into object

tiny-james committed:
add/2793: Moved test that selection is within container into assertion

tiny-james committed:
add/2793: Converted `Array#map` and `Array#filter` to lodash equalivents

tiny-james committed:
add/2793: Reorder calculations to avoid doing unneeded work

tiny-james committed:
add/2793: Break out of loop early to avoid excess indentation

tiny-james committed:
add/2793: Removed unused 'range' parameter of allowContext

tiny-james committed:
add/2793: Removed unneeded if-statment

tiny-james committed:
add/2793: Avoid some indentation with an early return

tiny-james committed:
add/2793: Replace escapeStringRegexp with escapeRegExp from lodash

tiny-james committed:
add/2793: Removed unneeded guard statement

tiny-james committed:
add/2793: Avoid spurious react reconcile by unwrapping function

tiny-james committed:
add/2793: Moved bounding client rect calculation into else branch

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

tiny-james committed:
add/2793: Only take padding into account when using the anchor

tiny-james committed:
Add/2793: Allow popover position to be overriden by prop getAnchorRect

tiny-james committed:
Merge branch 'master' into add/2793-user-autocompletion

EphoxJames commented:
@aduth I think I have addressed all your review points hence I am merging this.

youknowriad commented on components/popover/index.js:
```diff
@@ -42,6 +42,7 @@ class Popover extends Component {
 
```
I think the changes here may have broke `master`. All the popovers (inserter, menus...) are centered and not relative to their parent anymore

EphoxJames replied:
Yep, I see what you mean. I'm currently trying to figure out how my refactor changed that. Hopefully I'll have a fix for you by the time it rolls around to Europe's Monday.

EphoxJames replied:
Found the bug:
```
		return {
			...rect,
			top: rect.top + topPad,
			bottom: rect.bottom - bottomPad,
			height: rect.height - topPad - bottomPad,
		}
```
In the above `...rect` does not copy across the DomRect's properties as I had expected. The fix will thankfully be simple.

EphoxJames replied:
The bug should now be fixed.
